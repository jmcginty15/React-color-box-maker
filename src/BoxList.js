import { useState } from 'react';
import { v4 as uuid } from 'uuid';
import Box from './Box';
import NewBoxForm from './NewBoxForm';

const BoxList = () => {
    const [boxes, setBoxes] = useState([]);

    function handleNewBox({ height, width, backgroundColor }) {
        const newBoxes = [
            ...boxes,
            {
                height: height,
                width: width,
                backgroundColor: backgroundColor,
                id: uuid()
            }
        ]
        setBoxes(newBoxes);
    }

    function removeBox(evt) {
        const boxId = evt.target.id;
        const newBoxes = [...boxes];
        
        let i = 0;
        while (i < newBoxes.length && boxId !== newBoxes[i].id) i++;
        if (i !== newBoxes.length) newBoxes.splice(i, 1);
        setBoxes(newBoxes);
    }

    return (
        <div className="BoxList">
            {boxes.map(box => <Box key={box.id} id={box.id} height={box.height} width={box.width} backgroundColor={box.backgroundColor} handleClick={removeBox} />)}
            <NewBoxForm addBox={(formData) => handleNewBox(formData)} />
        </div>
    )
}

export default BoxList;