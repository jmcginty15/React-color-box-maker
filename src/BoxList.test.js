import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

// smoke test
it('renders without crashing', function () {
    render(<BoxList />);
});

// snapshot test
it('matches snapshot', function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it('adds a box on form submit', function () {
    const { queryByText } = render(<BoxList />);
    const submitButton = queryByText('Add Box');
    fireEvent.click(submitButton);
    const newBox = queryByText('X').parentElement;
    expect(newBox).toHaveStyle('height: 0');
    expect(newBox).toHaveStyle('width: 0');
    expect(newBox).toHaveStyle('background-color: #000000');
});

it('removes a box when remove button is clicked', function () {
    const { queryByText } = render(<BoxList />);
    const submitButton = queryByText('Add Box');
    fireEvent.click(submitButton);
    const newBoxRemoveButton = queryByText('X');
    const newBox = newBoxRemoveButton.parentElement;
    fireEvent.click(newBoxRemoveButton);
    expect(newBoxRemoveButton).not.toBeInTheDocument();
    expect(newBox).not.toBeInTheDocument();
});

it('displays and uses form inputs correctly', function () {
    const { queryByText, queryByTestId } = render(<BoxList />);
    const submitButton = queryByText('Add Box');

    const heightInput = queryByTestId('height-input');
    const widthInput = queryByTestId('width-input');
    const colorInput = queryByTestId('color-input');
    fireEvent.change(heightInput, { target: { value: 250 }});
    fireEvent.change(widthInput, { target: { value: 250 }});
    fireEvent.change(colorInput, { target: { value: '#555555' }});

    expect(heightInput.value).toBe("250");
    expect(widthInput.value).toBe("250");
    expect(colorInput.value).toBe("#555555");

    fireEvent.click(submitButton);
    const newBox = queryByText('X').parentElement;
    expect(newBox).toHaveStyle('height: 250px');
    expect(newBox).toHaveStyle('width: 250px');
    expect(newBox).toHaveStyle('background-color: #555555');

    expect(heightInput.value).toBe("0");
    expect(widthInput.value).toBe("0");
    expect(colorInput.value).toBe("#000000");
});