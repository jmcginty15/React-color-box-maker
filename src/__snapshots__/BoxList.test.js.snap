// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="NewBoxForm"
    >
      <h4
        class="NewBoxForm-header"
      >
        Add a New Box:
      </h4>
      <p
        class="NewBoxForm-line"
      >
        <label
          for="height"
        >
          Height in pixels:
        </label>
        <span
          class="NewBoxForm-input"
        >
          0 
          <input
            data-testid="height-input"
            max="500"
            min="0"
            name="height"
            type="range"
            value="0"
          />
        </span>
      </p>
      <p
        class="NewBoxForm-line"
      >
        <label
          for="width"
        >
          Width in pixels:
        </label>
        <span
          class="NewBoxForm-input"
        >
          0 
          <input
            data-testid="width-input"
            max="500"
            min="0"
            name="width"
            type="range"
            value="0"
          />
        </span>
      </p>
      <p
        class="NewBoxForm-line"
      >
        <label
          for="backgroundColor"
        >
          Background color:
        </label>
        <span
          class="NewBoxForm-input"
        >
          #000000 
          <input
            data-testid="color-input"
            name="backgroundColor"
            type="color"
            value="#000000"
          />
        </span>
      </p>
      <p
        class="NewBoxForm-line NewBoxForm-button"
      >
        <button
          type="submit"
        >
          Add Box
        </button>
      </p>
    </form>
  </div>
</DocumentFragment>
`;
